### PR TITLE
chore(dependencies): bump transformers v5

### PR DESCRIPTION
**Pull request by imstevenpmwork**
## Type / Scope

- **Type**: Chore | Dependencies | Refactor | CI
- **Scope**: Mostly policy code

## Summary / Motivation

- This makes the necessary changes to depend on `transformers>5.3.0`, consolidating on using a unique transformers version across all the policies that depend on it.

## Related issues

- Fixes / Closes: All the issues liked to the PR.
- Related: By doing this, we are also resolving the issue with the slow CI due to `uv` taking extended time to resolve the dependency tree (given that we had different transformers versions)

## What changed

- `pyproject.toml` now consolidates a consistent set of dependencies for all policies.
- Several fixes and improvements were included into the pi0 family as having a clear dependency tree allowed us to better test it and compare it to the original implementation.
- The base models in the HF Hub have been updated, tagging the previous versions to `v0.4` to avoid breaking current workflows (if the model is affected).
- When using the pi0 family policies after this PR, users will have to fine-tune again.

## How was this tested (or how to run locally)

* `pytest -qvs tests/`
* We also unblock some of our tests to be run in the CI that previously required accessing the Gated Models in the Hub. Increasing our testing surface in policies. This test mostly related to checking the logins of each policy to make sure no regression is introduced.
* Full tests CI run: https://github.com/huggingface/lerobot/actions/runs/22686669266
* E2E tests were also performed for LIBERO + making the checkpoint available for reproducibility. 

## Checklist (required before merge)

- [x] Linting/formatting run (`pre-commit run -a`)
- [x] All tests pass locally (`pytest`)
- [x] Documentation updated
- [x] CI is green

## Reviewer notes

Future work include:
- Tighter integration with transformers v5, when it comes to models & infrastructure
- Improve Pi0 performance
- Re-enable the test related to `resnet10`

### CODE REVIEW

File: src/lerobot/policies/pi0/modeling_pi0.py [hunk 1]
```diff
@@ -59,6 +59,12 @@ class ActionSelectKwargs(TypedDict, total=False):
     execution_horizon: int | None
 
 
+def _gated_residual(residual: torch.Tensor, hidden_states: torch.Tensor, gate: torch.Tensor) -> torch.Tensor:
```
Comment (imstevenpmwork) [inline src/lerobot/policies/pi0/modeling_pi0.py]: This was originally a private function in the custom transformers fork. Ported locally to remove the fork dependency.

File: src/lerobot/policies/pi0/modeling_pi0.py [hunk 2]
```diff
@@ -217,18 +223,53 @@ def resize_with_pad_torch(  # see openpi `resize_with_pad_torch` (exact copy)
     return padded_images
 
 
+class AdaRMSNorm(nn.Module):
```
Comment (imstevenpmwork) [inline src/lerobot/policies/pi0/modeling_pi0.py]: This replaces the custom fork's modified GemmaRMSNorm that accepted a `cond` parameter.

File: pyproject.toml [hunk 3]
```diff
@@ -146,7 +146,7 @@ groot = [
     "ninja>=1.11.1,<2.0.0",
     "flash-attn>=2.5.9,<3.0.0 ; sys_platform != 'darwin'"
 ]
-sarm = ["lerobot[transformers-dep]", "faker>=33.0.0,<35.0.0", "matplotlib>=3.10.3,<4.0.0", "qwen-vl-utils>=0.0.14,<0.1.0"]
+sarm = ["lerobot[transformers-dep]", "faker>=33.0.0,<35.0.0", "matplotlib>=3.10.3,<4.0.0", "qwen-vl-utils>=0.0.11,<0.1.0"]
```
Comment (imstevenpmwork) [inline pyproject.toml]: Changing this so `all` tag can include all policy tags. But yet to be tested if we can actually downgrade the minimum version

File: pyproject.toml [hunk 4]
```diff
@@ -127,13 +127,13 @@ phone = ["hebi-py>=2.8.0,<2.12.0", "teleop>=0.1.0,<0.2.0", "fastapi<1.0"]
...
-    "torchdiffeq==0.2.5",
-    "qwen_vl_utils==0.0.11"
+    "lerobot[transformers-dep]",
+    "peft>=0.18.0,<1.0.0",
+    "scipy==1.15.3", # TODO: Relax version
+    "torchdiffeq==0.2.5", # TODO: Relax version
+    "qwen-vl-utils==0.0.11" # TODO: Relax version
 ]
-pi = ["transformers @ git+https://github.com/huggingface/transformers.git@fix/lerobot_openpi", "scipy>=1.10.1,<1.15"]
+pi = ["lerobot[transformers-dep]", "scipy==1.15.3"] # TODO: Relax scipy version
```
Comment (imstevenpmwork) [inline pyproject.toml]: For now pinning `scipy` version to the same one used in the `wallx` tag (which is higher). Yet to be tested if this is possible, but ideally this shouldn't be pinned in any tag.

File: tests/policies/hilserl/test_modeling_classifier.py [hunk 5]
```diff
@@ -37,6 +38,9 @@ def test_classifier_output():
 
 
 @require_package("transformers")
+@pytest.mark.skip(
```
Comment (imstevenpmwork) [inline tests/policies/hilserl/test_modeling_classifier.py:41]: The custom `helper2424/resnet10` model on the Hub doesn't call `self.post_init()` in its `__init__`, which is required by transformers v5 to set `all_tied_weights_keys`. The `from_pretrained` call fails when it tries to access this attribute during `_finalize_model_loading`.

File: src/lerobot/policies/xvla/configuration_florence2.py [hunk 6]
```diff
@@ -276,6 +276,8 @@ def __init__(
         )
 
         # ensure backward compatibility for BART CNN models
+        if not hasattr(self, "forced_bos_token_id"):
+            self.forced_bos_token_id = None
```
Comment (imstevenpmwork) [inline src/lerobot/policies/xvla/configuration_florence2.py:280]: In transformers v5, `PretrainedConfig.__init__()` no longer sets `forced_bos_token_id` as a default attribute (generation params moved to `GenerationConfig`). The vendored Florence2LanguageConfig accessed `self.forced_bos_token_id` after `super().__init__()` but it was never set.

File: src/lerobot/policies/pi05/modeling_pi05.py [hunk 7]
```diff
@@ -215,18 +220,53 @@ def resize_with_pad_torch(  # see openpi `resize_with_pad_torch` (exact copy)
     return padded_images
 
 
+class AdaRMSNorm(nn.Module):
```
Comment (jadechoghari) [inline src/lerobot/policies/pi05/modeling_pi05.py]: Do you use this class somewhere in the code? other than the type annotations?

File: src/lerobot/policies/pi0/modeling_pi0.py [hunk 8]
```diff
@@ -217,18 +222,53 @@ def resize_with_pad_torch(  # see openpi `resize_with_pad_torch` (exact copy)
...
-    models = [paligemma.language_model, gemma_expert.model]
+    models = [paligemma.model.language_model, gemma_expert.model]
     query_states = []
     key_states = []
     value_states = []
     gates = []
     for i, hidden_states in enumerate(inputs_embeds):
         layer = models[i].layers[layer_idx]
-        hidden_states, gate = layer.input_layernorm(hidden_states, cond=adarms_cond[i])  # noqa: PLW2901
+        if isinstance(layer.input_layernorm, AdaRMSNorm):
```
Comment (jadechoghari) [inline src/lerobot/policies/pi0/modeling_pi0.py]: In the new version of transformer v5, this is always false since the pinned version contain the new special layer norm AdaRMSNorm in gemma we need.
We should probably do:
```python
self.input_layernorm = GemmaRMSNorm(config.hidden_size, eps=config.rms_norm_eps, cond_dim=cond_dim)
```
And remove this conditional check

File: src/lerobot/policies/pi0/modeling_pi0.py [hunk 9]
```diff
@@ -277,15 +317,19 @@ def compute_layer_complete(
...
         # first residual
-        out_emb = modeling_gemma._gated_residual(hidden_states, out_emb, gates[i])  # noqa: SLF001
+        out_emb = _gated_residual(hidden_states, out_emb, gates[i])  # noqa: SLF001
         after_first_residual = out_emb.clone()
-        out_emb, gate = layer.post_attention_layernorm(out_emb, cond=adarms_cond[i])
+        if isinstance(layer.post_attention_layernorm, AdaRMSNorm):
+            out_emb, gate = layer.post_attention_layernorm(out_emb, cond=adarms_cond[i])
+        else:
+            out_emb = layer.post_attention_layernorm(out_emb)
+            gate = torch.ones(out_emb.shape[:-1], dtype=out_emb.dtype, device=out_emb.device)
```
Comment (jadechoghari) [inline src/lerobot/policies/pi0/modeling_pi0.py]: Same thing her,e we should remove this conditional check and do
```python
self.post_attention_layernorm = GemmaRMSNorm(config.hidden_size, eps=config.rms_norm_eps, cond_dim=cond_dim)
```
the post-attention layernorm is a modified version and differs from the current Gemma implementation.

File: src/lerobot/policies/pi0/modeling_pi0.py [hunk 10]
```diff
@@ -217,18 +222,53 @@ def resize_with_pad_torch(  # see openpi `resize_with_pad_torch` (exact copy)
     return padded_images
 
 
+class AdaRMSNorm(nn.Module):
```
Comment (pkooij) [inline src/lerobot/policies/pi0/modeling_pi0.py]: Only Pi0.5 uses AdaRMSNorm, so maybe we can just exclude it entirely in pi0? See `use_adarms`

File: src/lerobot/policies/pi0/modeling_pi0.py (hunk 8 above)
Comment (imstevenpmwork) [reply src/lerobot/policies/pi0/modeling_pi0.py]: Addressed in: https://github.com/huggingface/lerobot/pull/2967

File: src/lerobot/policies/pi0/modeling_pi0.py (hunk 9 above)
Comment (imstevenpmwork) [reply src/lerobot/policies/pi0/modeling_pi0.py]: Addressed in: https://github.com/huggingface/lerobot/pull/2967

File: src/lerobot/policies/pi0/modeling_pi0.py (hunk 10 above)
Comment (imstevenpmwork) [reply src/lerobot/policies/pi0/modeling_pi0.py]: Addressed in: https://github.com/huggingface/lerobot/pull/2967

File: tests/policies/pi0_fast/test_pi0_fast_original_vs_lerobot.py [hunk 11]
```diff
@@ -54,19 +54,19 @@
 IMAGE_WIDTH = 224
 NUM_VIEWS = 2  # Number of camera views
 DEVICE = "cuda" if torch.cuda.is_available() else "cpu"
-MODEL_PATH_LEROBOT = "lerobot/pi0fast-base"
+MODEL_PATH_LEROBOT = "jadechoghari/pi0fast-base"
```
Comment (jadechoghari) [inline tests/policies/pi0_fast/test_pi0_fast_original_vs_lerobot.py]: before merging, we should update pifast base in lerobot and undo this

File: pyproject.toml [hunk 12]
```diff
@@ -129,13 +129,13 @@ phone = ["hebi-py>=2.8.0,<2.12.0", "teleop>=0.1.0,<0.2.0", "fastapi<1.0"]
...
-    "transformers==4.49.0",
-    "peft==0.17.1",
-    "scipy==1.15.3",
-    "torchdiffeq==0.2.5",
-    "qwen_vl_utils==0.0.11"
+    "lerobot[transformers-dep]",
+    "peft>=0.18.0,<1.0.0",
+    "scipy==1.15.3", # TODO: Relax version
+    "torchdiffeq==0.2.5", # TODO: Relax version
+    "qwen-vl-utils==0.0.11" # TODO: Relax version
```
Comment (imstevenpmwork) [inline pyproject.toml]: Can these versions be relaxed? @pkooij

File: pyproject.toml (hunk 3 above)
Comment (imstevenpmwork) [reply pyproject.toml]: @pkooij Would this pose an issue ?

File: pyproject.toml (hunk 4 above)
Comment (imstevenpmwork) [reply pyproject.toml]: @pkooij would this pose an issue ? Ideally we should relax this dep

File: pyproject.toml (hunk 3 above)
Comment (pkooij) [reply pyproject.toml]: no I don't think so

File: pyproject.toml (hunk 4 above)
Comment (pkooij) [reply pyproject.toml]: hard to know for sure, only by running tests

File: pyproject.toml (hunk 12 above)
Comment (pkooij) [reply pyproject.toml]: See anwsers above

File: pyproject.toml (hunk 4 above)
Comment (jadechoghari) [reply pyproject.toml]: imo it makes sense to train/eval on one of our simulation benchmarks to get the answer

File: pyproject.toml (hunk 3 above)
Comment (imstevenpmwork) [reply pyproject.toml]: Addressed in: https://github.com/huggingface/lerobot/pull/3034

File: pyproject.toml (hunk 4 above)
Comment (imstevenpmwork) [reply pyproject.toml]: Addressed in: https://github.com/huggingface/lerobot/pull/3034

File: pyproject.toml (hunk 12 above)
Comment (imstevenpmwork) [reply pyproject.toml]: Addressed in: https://github.com/huggingface/lerobot/pull/3034

**Comment by imstevenpmwork** [comment]
Today's Full CI: https://github.com/huggingface/lerobot/actions/runs/22455225460

**Comment by imstevenpmwork** [comment]
Today's CI Full Test results: https://github.com/huggingface/lerobot/actions/runs/22589307815
The only error should be related to: https://github.com/huggingface/lerobot/pull/3050

**Comment by imstevenpmwork** [comment]
Transformers has release the patch release that was blocking our CI, now that we have updated it our Full CI test from today should all be green: https://github.com/huggingface/lerobot/actions/runs/22686669266

File: tests/policies/hilserl/test_modeling_classifier.py (hunk 5 above)
Comment (imstevenpmwork) [reply tests/policies/hilserl/test_modeling_classifier.py:41]: @helper2424 Just notifying you that your models might need some update after transformers v5 lands on `main` (soon)

File: tests/policies/pi0_fast/test_pi0_fast_original_vs_lerobot.py (hunk 11 above)
Comment (imstevenpmwork) [reply tests/policies/pi0_fast/test_pi0_fast_original_vs_lerobot.py]: Addressed in commit: https://github.com/huggingface/lerobot/pull/2964/changes/ab7dcc2f792319d720d190c23d713ca2b2d9fa96

**Review by pkooij — DISMISSED** [review]
Let go

**Review by jadechoghari — APPROVED** [review]
This was a great effort, super happy to see the Transformers v5 upgrade land after all the work here. LGTM 🎉
We've updated all our LeRobot checkpoints to reflect the upgraded changes. The affected checkpoints are now labeled with an extra v0.4. If a checkpoint does not have this label, then the model is not affected by this upgrade. If you trained on pi0 / pi0fast / pi05, you will likely need to fine-tune again for transformers v5, as we changed some part of the processing/modeling (for better reproducibility!!)

What's left for the next patch:
- add weight tying instead of manual tying

Our evaluation was conducted on the Libero benchmark. We provide the following checkpoint trained on Libero for you to try with this upgrade:
Pi0Fast: https://huggingface.co/lerobot/pi0fast-libero
Pi05: https://huggingface.co/lerobot/pi05-libero

File: src/lerobot/policies/pi05/modeling_pi05.py [hunk 13]
```diff
@@ -880,6 +887,7 @@ def denoise_step(
         full_att_2d_masks_4d = self._prepare_attention_masks_4d(full_att_2d_masks)
         self.paligemma_with_expert.gemma_expert.model.config._attn_implementation = "eager"  # noqa: SLF001
 
+        past_key_values = copy.deepcopy(past_key_values)
```
Comment (mgehre-amd) [inline src/lerobot/policies/pi05/modeling_pi05.py:890]: @imstevenpmwork 
This seems to negatively affect performance by causing a graph break in torch.compile. Why is the deepcopy needed here? The only use is in `paligemma_with_expert.forward` where we pass `use_cache=False`, which should keep the cache from being modified.
